### PR TITLE
8306992: [JVMCI] mitigate more against JVMCI related OOME causing VM to exit

### PR DESCRIPTION
**Pull request by dougxc**
This PR makes the following changes to mitigate against an OOME or other recoverable error in JVMCI from causing the VM to exit:
* Tracks upcalls into libjvmci or creation of libjvmci.
* If 10% or more of these calls fail with an uncaught exception, then JVMCI compilation is disabled (i.e. future compilations fall back to Tier 1).

When JVMCI compilation is disabled, a warning is emitted:
```
[0.064s][warning][jit,compilation] JVMCI compiler disabled after 11 of 15 upcalls had errors (Last error: "uncaught exception in call_HotSpotJVMCIRuntime_compileMethod"). Use -Xlog:jit+compilation for more detail.
```

With `-Xlog:jit+compilation`, the extra detail shown is:
```
[0.182s][info][jit,compilation] uncaught exception in call_HotSpotJVMCIRuntime_compileMethod while compiling java.util.stream.StreamOpFlag.fromCharacteristics(Ljava/util/Spliterator;)I
Exception in thread "JVMCI-native CompilerThread0": java.lang.InternalError
java.lang.InternalError: aborting compilation of HotSpotMethod<Object.<init>()>
	at jdk.internal.vm.ci@20.0.2-internal/jdk.vm.ci.hotspot.HotSpotJVMCIRuntime.compileMethod(HotSpotJVMCIRuntime.java:923)
```

Note that the errors treated by the changes in the PR are expected to be exceedingly rare. For example, an OOME while starting a libgraal isolate or initializing the JVMCI compiler. Exceptions thrown during compilation are already handled by the Graal [CompilationWrapper](https://github.com/oracle/graal/blob/431ecf7d26f56cee49708854fe0e89b05514492b/compiler/src/jdk.internal.vm.compiler/src/org/graalvm/compiler/core/CompilationWrapper.java#L65).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306992](https://bugs.openjdk.org/browse/JDK-8306992): [JVMCI] mitigate more against JVMCI related OOME causing VM to exit

### Reviewers
 * [Tom Rodriguez](https://openjdk.org/census#never) (@tkrodriguez - **Reviewer**) ⚠️ Review applies to [26a1c426](https://git.openjdk.org/jdk/pull/13905/files/26a1c426e0fd64c43a023a6f942fc5aca987ba24)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13905/head:pull/13905` \
`$ git checkout pull/13905`

Update a local copy of the PR: \
`$ git checkout pull/13905` \
`$ git pull https://git.openjdk.org/jdk.git pull/13905/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13905`

View PR using the GUI difftool: \
`$ git pr show -t 13905`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13905.diff">https://git.openjdk.org/jdk/pull/13905.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13905#issuecomment-1542323779)

### CODE REVIEW

**Review by tkrodriguez — APPROVED** [review]

**Comment by dougxc** [comment]
I rebased this PR to remove commits from https://github.com/openjdk/jdk/pull/14000 that accidentally got cherry-picked into this PR.

**Comment by dougxc** [comment]
@tkrodriguez could you please review https://github.com/openjdk/jdk/pull/13905/commits/fb2250921fc731a47a945a3e578b384a6849c331 which I added to handle cases such as https://github.com/oracle/graal/discussions/6216#discussioncomment-5942123.

File: src/hotspot/share/jvmci/jvmciRuntime.cpp [hunk 1]
```diff
@@ -1225,10 +1226,16 @@ bool JVMCIRuntime::detach_thread(JavaThread* thread, const char* reason, bool ca
   return destroyed_javavm;
 }
 
-JNIEnv* JVMCIRuntime::init_shared_library_javavm() {
+JNIEnv* JVMCIRuntime::init_shared_library_javavm(int* create_JavaVM_err) {
   MutexLocker locker(_lock);
   JavaVM* javaVM = _shared_library_javavm;
   if (javaVM == nullptr) {
+    const char* val = Arguments::PropertyList_get_value(Arguments::system_properties(), "test.jvmci.forceEnomemOnLibjvmciInit");
```
Comment (dougxc) [inline src/hotspot/share/jvmci/jvmciRuntime.cpp:1233]: Note that this view on system properties is restricted to properties set at VM startup (e.g. on the command line) and will not see the result of calls to `System.setProperty()` made by an application.

**Comment by dougxc** [comment]
Thanks for the reviews @tkrodriguez .

/integrate

**Pull request closed without merging**